hooks/_common.sh

```diff
@@ -170,6 +170,68 @@ function common::is_hook_run_on_whole_repo {
   fi
 }
 
+#######################################################################
+# Get the number of CPU logical cores available for pre-commit to use
+# Arguments:
+#  parallelism_ci_cpu_cores (string) Used in edge cases when number of
+#    CPU cores can't be derived automatically
+# Outputs:
+#   Returns number of CPU logical cores, rounded down to nearest integer
+#######################################################################
+function common::get_cpu_num {
+  local -r parallelism_ci_cpu_cores=$1
+
+  local millicpu
+
+  if [[ -f /sys/fs/cgroup/cpu/cpu.cfs_quota_us ]]; then
+    # Inside K8s pod or DinD in K8s
+    millicpu=$(< /sys/fs/cgroup/cpu/cpu.cfs_quota_us)
+
+    if [[ $millicpu -eq -1 ]]; then
+      # K8s no limits or in DinD
+      if [[ -n $parallelism_ci_cpu_cores ]]; then
+        # 22 EINVAL Invalid argument. Some invalid argument was supplied.
+        [[ $parallelism_ci_cpu_cores =~ ^[[:digit:]]+$ ]] || return 22
+
+        echo "$parallelism_ci_cpu_cores"
+        return
+      fi
+
+      common::colorify "yellow" "Unable to derive number of available CPU cores.\n" \
+        "Running inside K8s pod without limits or inside DinD without limits propagation.\n" \
+        "To avoid possible harm, parallelism is disabled.\n" \
+        "To re-enable it, set corresponding limits, or set the following for the current hook:\n" \
+        "  args:\n" \
+        "    - --hook-config=--parallelism-ci-cpu-cores=N\n" \
+        "where N is the number of CPU cores to allocate to pre-commit."
+
+      echo 1
+      return
+    fi
+
+    echo $((millicpu / 1000))
+    return
+  fi
+
+  if [[ -f /sys/fs/cgroup/cpu.max ]]; then
+    # Inside Linux (Docker?) container
+    millicpu=$(cut -d' ' -f1 /sys/fs/cgroup/cpu.max)
+
+    if [[ $millicpu == max ]]; then
+      # No limits
+      nproc 2> /dev/null || echo 1
+      return
+    fi
+
+    echo $((millicpu / 1000))
+    return
+  fi
+
+  # On host machine or any other case
+  # `nproc` - linux, `sysctl -n hw.ncpu` - macOS, `echo 1` - fallback
```

Comment:
```suggestion
  # `nproc` - linux, `sysctl -n hw.ncpu` - BSD/macOS, `echo 1` - fallback
```

Reply:
Maybe then
```suggestion
  # `nproc` - Linux/BSD, `sysctl -n hw.ncpu` - macOS/BSD, `echo 1` - fallback
```
?

https://man.freebsd.org/cgi/man.cgi?query=nproc&sektion=1&format=html

Reply:
Yep, makes sense (my FreeBSD VM is quite outdated and hence it's missing a lot of tools including the `nproc` 😿). As for the `BSD/macOS` vs `macOS/BSD`: since macOS is the, hmm, derivative of BSD, I feel like `BSD` should go first =) Though this is not essential obviously =))

Reply:
Up to you